### ISSUE TITLE
Generic components and tags types are not supported and throw exception on usage.

### ISSUE THREAD

**Issue by friflo** [friflo/Friflo.Json.Fliox#53]
The issue came up in: https://github.com/friflo/Friflo.Json.Fliox/issues/51#issuecomment-2176902166

Generic component and tags type are not supported in library version <= 2.0.0.
E.g. 

```cs
public struct GenericComponent<T> : IComponent {
    public T Value;
}

public struct GenericTag<T> : ITag { }
```

These types are not registered as component / tag type and throw an exception when using them.
```
System.TypeInitializationException : The type initializer for 'Friflo.Engine.ECS.TagInfo`1' threw an exception.
  ----> System.Collections.Generic.KeyNotFoundException : The given key 'Tests.ECS.GenericTag`1[System.Int32]' was not present in the dictionary.
   at Friflo.Engine.ECS.Tags.Get[T]() 
   at Friflo.Engine.ECS.Entity.AddTag[TTag]()
   at Tests.ECS.Base.Test_Components.Test_Generic_Component()
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
--KeyNotFoundException
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Friflo.Engine.ECS.SchemaTypeUtils.GetTagIndex(Type type)
   at Friflo.Engine.ECS.TagInfo`1..cctor()
```

**Comment by friflo** [comment]
My plan to support generic component / tag types:
An attribute needs to be added to those types. E,g,

```cs
[GenericInstanceType(typeof(int))]
public struct GenericComponent<T> : IComponent {
    public T Value;
}

[GenericInstanceType(typeof(int))]
[GenericInstanceType(typeof(string))]
public struct GenericTag<T> : ITag { }
```

I guess I will finish this today an publish a nuget package.

**Comment by friflo** [comment]
Just published package 2.1.0 supporting Generic components and tags types on nuget.
These type must be annotated as shown in the example above.

In case of a missing attribute an exception is thrown like:
```
System.TypeInitializationException : The type initializer for 'Friflo.Engine.ECS.TagInfo`1' threw an exception.
  ----> System.InvalidOperationException : Missing attribute [GenericInstanceType("<key>", typeof(Int32), typeof(String))] for generic ITag type: Tests.ECS.GenericTag2`2[System.Int32,System.String]
   at Friflo.Engine.ECS.Tags.Get[T]() in G:\_dev\Fliox.Engine\Engine\src\ECS\Archetype\Tags.cs:line 146
   at Friflo.Engine.ECS.Entity.AddTag[TTag]() in G:\_dev\Fliox.Engine\Engine\src\ECS\Entity.cs:line 428
   at Tests.ECS.Base.Test_Components.Test_Generic_Component_exceptions() in G:\_dev\Fliox.Engine\Engine\src\Tests\ECS\Base\Test_Components.cs:line 71
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
--InvalidOperationException
   at Friflo.Engine.ECS.SchemaTypeUtils.GetTagIndex(Type type) in G:\_dev\Fliox.Engine\Engine\src\ECS\Base\SchemaTypeUtils.cs:line 27
   at Friflo.Engine.ECS.TagInfo`1..cctor() in G:\_dev\Fliox.Engine\Engine\src\ECS\Base\Types\TagType.cs:line 36
```

The inner exception describes how to fix the issue:
```
System.InvalidOperationException : Missing attribute [GenericInstanceType("<key>", typeof(Int32), typeof(String))] for generic ITag type: Tests.ECS.GenericTag2`2[System.Int32,System.String]
```

**Comment by friflo** [comment]
**Note**
I would not advocate to use generic component & tag types.
Component & tag types should be as simple as possible.
Making them generic adds a layer of abstraction. They are more complicated in usage and understanding.

Beside this their performance and memory characteristics are same as non generic types.
